src/toil/batchSystems/slurm.py

```diff
@@ -444,6 +443,29 @@ def prepareSbatch(self,
             if cpu is not None:
                 sbatch_line.append(f'--cpus-per-task={math.ceil(cpu)}')
 
+            if gpus:
+                # This block will add a gpu supported partition only if no partition is supplied by the user
+                sbatch_line = sbatch_line[:1] + [f'--gres=gpu:{gpus}'] + sbatch_line[1:]
+                lowest_priority_gpus = sorted([i for i in self.boss.partitions if i["gres"]], key=lambda x: x["priority"])
```

Comment:
I would move this to happen at startup and split it into two things: a default GPU partition that we select once (or leave None if we don't have any partitions that seem to have GPUs) and a set of partitions we think have GPUs (for fast membership testing for the warning). I don't think "a list of all the GPU partitions sorted so the default one is first" is quite the right representation.